File: backend/app/api/notifications.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select, func, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.security import get_current_user
from app.models.notification import Notification
from app.models.user import User
from app.schemas import NotificationOut, NotificationListOut

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


@router.get("", response_model=NotificationListOut)
async def list_notifications(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """List current user's notifications (latest first) with unread count."""
    base = select(Notification).where(Notification.user_id == current_user.id)

    total_result = await db.execute(
        select(func.count()).select_from(base.subquery())
    )
    total = total_result.scalar() or 0

    unread_result = await db.execute(
        select(func.count(Notification.id)).where(
            Notification.user_id == current_user.id,
            Notification.is_read.is_(False),
        )
    )
    unread_count = unread_result.scalar() or 0

    result = await db.execute(
        base.order_by(Notification.created_at.desc(), Notification.id.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
    )
    notifications = result.scalars().all()
    return NotificationListOut(
        items=[NotificationOut.model_validate(n) for n in notifications],
        unread_count=unread_count,
        total=total,
        page=page,
        page_size=page_size,
    )


@router.get("/unread-count")
async def get_unread_count(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Lightweight endpoint for polling the unread badge."""
    result = await db.execute(
        select(func.count(Notification.id)).where(
            Notification.user_id == current_user.id,
            Notification.is_read.is_(False),
        )
    )
    return {"unread_count": result.scalar() or 0}


@router.post("/read-all")
async def mark_all_notifications_read(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Mark all of the current user's notifications as read."""
    result = await db.execute(
        update(Notification)
        .where(
            Notification.user_id == current_user.id,
            Notification.is_read.is_(False),
        )
        .values(is_read=True)
    )
    await db.flush()
    return {"message": "全部已读", "updated": result.rowcount or 0}


@router.post("/{notification_id}/read", response_model=NotificationOut)
async def mark_notification_read(
    notification_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Mark a single notification as read."""
    result = await db.execute(
        select(Notification).where(
            Notification.id == notification_id,
            Notification.user_id == current_user.id,
        )
    )
    notification = result.scalar_one_or_none()
    if not notification:
        raise HTTPException(status_code=404, detail="通知不存在")
    notification.is_read = True
    await db.flush()
    await db.refresh(notification)
    return NotificationOut.model_validate(notification)

File: frontend/src/api/notifications.ts
import api from '../utils/api'

export type NotificationType =
  | 'task_assigned'
  | 'comment'
  | 'mention'
  | 'user_approved'
  | 'user_rejected'
  | 'member_added'
  | string

export interface AppNotification {
  id: number
  user_id: number
  type: NotificationType
  title: string
  body: string
  link: string
  is_read: boolean
  created_at: string
}

export interface NotificationListResponse {
  items: AppNotification[]
  unread_count: number
  total: number
  page: number
  page_size: number
}

export async function fetchNotifications(page = 1, pageSize = 20): Promise<NotificationListResponse> {
  const res = await api.get('/notifications', { params: { page, page_size: pageSize } })
  return res.data
}

export async function fetchUnreadCount(): Promise<number> {
  const res = await api.get('/notifications/unread-count')
  return res.data.unread_count as number
}

export async function markNotificationRead(id: number): Promise<AppNotification> {
  const res = await api.post(`/notifications/${id}/read`)
  return res.data
}

export async function markAllNotificationsRead(): Promise<void> {
  await api.post('/notifications/read-all')
}
